# Understanding different Pytorch tensor functions

### Pytorch library

Pytorch is an open-source machine learning library that has been developed by Facebook's AI Research lab (FAIR). It is mostly used for developing algorithms related to machine learning and deep learning. Additionally, this library can perform operations related to concepts of linear algebra and some optimization problem methods.

### Pytorch Tensor
The fundamental unit of this library is the tensor data structure. 
The tensor data structure is a multidimensional array containing a particular datatype as it's input. It is the basic building block for all the functions described in PyTorch. 
In this notebook, we discuss five different PyTorch tensor functions that are used to modify tensors to get the required outputs. They are:
* The norm of a tensor 
* Dot product of a tensor
* The inverse of a tensor, 
* Selecting index from a tensor 
* Cosine similarity of two tensors: The function cosine similarity is generated using function 1 which is the tensor norm and function 2 which is a tensor dot product.

#### Tensor.norm
Norm function gives information about the distance of a particular vector from the origin. Additionally, it provides information about the distance between the two vectors in an N-dimensional space, often referred to as the euclidean norm.

The formula for the vector norm with 3 elements is defined as:

<img src="vector_norm_with_3_dim.png">

The formula for the matrix norm referred to as Frobenius norm is defined as:

<img src="Forbenious_matrix_norm.png">

Norm functions are vastly used in the machine learning algorithm's development. Few of the examples are described below:

* They are used as regularizer functions, mean square error functions in regression problems. 
* In the KNN algorithm norm function is used to measure the distance between two vectors.
* Loss functions such as mean square errors and mean absolute error.

#### Tensor.dotproduct
Computes the dot product between two tensors. 

It gives the information related to the orientation of each of the vector and when superimposed over one another how the magnitude is increased or decreased in the resulting vector based on their initial orientation.

The formula for dot product is:

<img src="Dot_product_of_two_vectors.png">


It is extensively used in machine applications, a few of them are described below:
* In NLP applications it is used to understand how two documents are similar by computing the dot product of their document word vectors
* It forms the fundamental unit for computing the output of the neurons in machine learning and deep learning applications.

####  Tensor.inverse

Computes the inverse of the square matrix input. Input can be batches of 2D square tensors, in which case this function would return a tensor composed of individual inverses. 
Gauss Jordan matrix elimination method and Cramer's rule method are some of the methods by which the inverse of the matrices are found.
 
Applications of the torch. inverse function in machine learning is described below:
* It is used in the linear regression algorithm inverse can help us to compute resultant weight vector or referred to as regression coefficients in the matrix form. 
Let
* X -> be the input data (training data)
* X' -> Transpose of the input data
* Y -> be the output data (variables we are trying to predict)
* B -> Weight vector(the slope coefficients or coefficients of the linear model) 

* The weight vector is calulated as :

<img src="Linear_regression_weight_vector_calculation.png">




#### Tensor.select_index

Index select functions are possibly used for selecting a particular range of data to understand and get more insights about the data. This function can be used to generate required batch sizes while training the data in machine learning and deep learning applications.

#### Tesnor.consine_similarity

Cosine similarity measures the similarity between two vectors. It is measured by the cosine of the angle between two vectors and determines whether two vectors are pointing in roughly the same direction. In finding out the similarity, firstly we calculate the norm corresponding to the two inputs provided, and then further on calculating the dot product. Following formula gives the information about the cosine similarity:


<img src="Cosine_similarity_formula.png">


cosine(angle) gives the similarity score between the two vectors. The higher the score more similar are the vectors.

#### Applications of cosine similarity functions 

In natural language processing cosine similarity can be used to find the similarity between the two documents utilizing the word document vector. 

For the feature vector in machine learning and deep learning problems if we find that two feature vectors are very similar, but with a slight difference, then we can also use it to reduce the dimensionality of the feature space.

In [2]:
# Import torch and other required modules
import torch
import numpy

## Function 1 - Torch.norm
This PyTorch function returns a vector norm or matrix norm based on the tensor input provided(i.e. vector or matrix).

Vector norm - Norm of a vector is used to describe the magnitude of an N-dimensional vector. Norm intuitively gives information about the distance of the vector from the origin.

Matrix norm -  Forbenious norm is used to calculate the norm of the matrix, but there are other kinds of norms such as maximum absolute column sum norm, spectral norm, and maximum absolute column row norm.

The following examples give a method of calculating norm using torch tensors

* Example 1: We calculate the norm of a vector using the euclidean norm method.
* Example 2: We calculate the norm of a matrix using the Frobenius norm

This function takes in the following arguments:

        input (Tensor): the input tensor
        p (int, float, inf, -inf, 'fro', 'nuc', optional): the order of norm. Default: fro
        The following norms can be calculated:

           =====  ============================  ==========================
            ord    matrix norm                   vector norm
            =====  ============================  ==========================
            None   Frobenius norm                2-norm
            'fro'  Frobenius norm                --
            'nuc'  nuclear norm                  --
            Other  as vec norm when dim is None  sum(abs(x)**ord)**(1./ord)
            =====  ============================  ==========================

        dim (int, 2-tuple of ints, 2-list of ints, optional): If it is an int,
            vector norm will be calculated, if it is 2-tuple of ints, matrix norm
            will be calculated. If the value is None, matrix norm will be calculated
            when the input tensor only has two dimensions, vector norm will be
            calculated when the input tensor only has one dimension. If the input
            tensor has more than two dimensions, the vector norm will be applied to
            last dimension.
        keepdim (bool, optional): whether the output tensors have :attr:`dim`
            retained or not. Ignored if :attr:`dim` = ``None`` and
            :attr:`out` = ``None``. Default: ``False``
        out (Tensor, optional): the output tensor. Ignored if
            :attr:`dim` = ``None`` and :attr:`out` = ``None``.
        dtype (:class:`torch.dtype`, optional): the desired data type of
            returned tensor. If specified, the input tensor is casted to
            :attr:'dtype' while performing the operation. Default: None.

In [172]:
# Example 1 - norm of a vector
import numpy as np

def vector_norm_caluclator(input_numpy_array):
    input_tensor_vector = np.array(input_numpy_array,dtype=np.float64)
    torch_tesnor_vector = torch.from_numpy(input_tensor_vector)
    torch_norm_output = torch.norm(torch.tensor(torch_tesnor_vector,dtype=torch.float),dim=0)
    return torch_norm_output.numpy()

input_numpy_array = [1,2,3,4,5]
output = vector_norm_caluclator(input_numpy_array)
print ("The vector norm is:"+str(output))

The vector norm is:7.4161983


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:7: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  import sys


In this example, we give a 5 elements vector array as the input, which is provided through the NumPy library and it's converted to torch tensor array using PyTorch library. Norm is calculated is torch.norm function and finally, convert to NumPy array and return the norm output which will be 1 dimension output.

For this example : 
* input = [1,2,3,4,5] 
* output = The vector norm is:7.416198487095663

In [173]:
# Example 2 - norm of a matrix
import numpy as np

def matrix_norm_caluclator(input_numpy_array):
    input_tensor_vector = np.matrix(input_numpy_array,dtype=np.float64)
    torch_tesnor_vector = torch.from_numpy(input_tensor_vector)
    torch_norm_output = torch.norm(torch.tensor(torch_tesnor_vector,dtype=torch.float),p='fro',dim=1)
    return torch_norm_output.numpy()

input_numpy_array = [[1.,2.],[3.,4.]]

output = matrix_norm_caluclator(input_numpy_array)
print ("The matrix norm is:"+str(output))


The matrix norm is:[2.236068 5.      ]


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:7: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  import sys


The matrix norm in our case is called the Frobenius norm. In this example, we give a (2,2) dimensional matrix as input through a NumPy matrix and it's converted to torch tensor array. The torch.norm function is used to calculate the norm of the matrix and is converted to a NumPy array. 

For this example : input = [[1,2],[3,4]] 


*  `matrix    col0   col1`

* ` row0       1      2`

* ` row1       3      4`


a) Since the dim=0 the output of the norm will be a row vector as the norm will be calculated across each column:

* col0 : $squareroot(1^2+3^2)$ = 3.16

* col1 : $squareroot(2^2+4^2)$ = 4.47

* The matrix norm is : [3.16, 4.47]

b) Since the dim=1 the output of the norm will be a row vector as the norm will be calculated across each column:

* row0 : $squareroot(1^2+2^2)$ = 2.23

* row1 : $squareroot(3^2+4^2)$ = 5

* The matrix norm is : [3.16, 4.47]

c) with default dim=None, the Frobenius norm is calculated which Square root of the sum of all elements:

* The norm output is : $sqrt(1^2+2^2+3^2+4^2)$ = 7.416198487095663



In [38]:
# Example 3 - breaking (to illustrate when it breaks)
def vector_norm_caluclator(input_numpy_array):
    input_tensor_vector = np.array(input_numpy_array,dtype=np.float64)
    print (np.shape(input_tensor_vector))
    torch_tesnor_vector = torch.from_numpy(input_tensor_vector)
    torch_norm_output = torch.norm(torch.tensor(torch_tesnor_vector),p='fro',dim=1)
    return torch_norm_output.numpy()

input_numpy_array = [1,2,3,4,5]
output = vector_norm_caluclator(input_numpy_array)
print ("The vector norm is:"+str(output))

(5,)


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:6: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  


IndexError: Dimension out of range (expected to be in range of [-1, 0], but got 1)

Here we provide the same input to the system but we change the default dimension parameter which is `dim` while calculating the norm. Since we pass the input as a 1d array with the 5 elements, the output for `dim=0` parameter will be calculated by computing the norm across all the 1d array elements. If the `dim=1` since no elements are corresponding to the other dimension the output throws an error, that the dimension out of range.





#### Applications of norm function : 
This function is used to calculate the norm of the vector, it's typically useful while calculating the distance of the vector from origin or from another vector. In machine learning, L1 and L2 norms are typically used as regularizers in regression problems. They are used to calculate the distance between two vectors like in mean square error calculation( the difference between the actual vector and output vector) which is also a performance measure in linear regression. Norm is applied to various different concepts in machine learning and its a quite essential function.

##  Function 2 - Torch.dot
This PyTorch function Computes the dot product (inner product) of two tensors.

torch.dot function is evaluated as follows:

* Example1: 
* a = [1.,2.] b = [3.,4.]
  dot_prod = $1*3+2*4$ = 11.0
* Example 2:
* a = [1.,2.,5.] b = [3.,4.,10.]
  dot_prod = $1*3+2*4+5*10$ = 61.0

In [3]:
# Example 1 - working
import numpy as np

def dot_product_calculator(input_numpy_array1, input_numpy_arry2):
    input_tensor_vector1 = np.array(input_numpy_array1,dtype=np.float64)
    input_tensor_vector2 = np.array(input_numpy_array2,dtype=np.float64)
    torch_tesnor_vector1 = torch.from_numpy(input_tensor_vector1)
    torch_tesnor_vector2 = torch.from_numpy(input_tensor_vector2)
    torch_norm_output = torch.dot(torch.tensor(torch_tesnor_vector1),torch.tensor(torch_tesnor_vector2))
    return torch_norm_output.numpy()

input_numpy_array1 = [1.,2.]
input_numpy_array2 = [3.,4.]
output = dot_product_calculator(input_numpy_array1,input_numpy_array2)
print ("The dot product is:"+str(output))


The dot product is:11.0


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:9: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  if __name__ == '__main__':


The dot product is computed with the input arrays as 
input_1 = [1,2]
input_2 = [3,4]

output is = 1 * 3 + 2 * 4 = 11


In [175]:
# Example 2 - working
import numpy as np

def dot_product_calculator(input_numpy_array1, input_numpy_arry2):
    input_tensor_vector1 = np.array(input_numpy_array1,dtype=np.float64)
    input_tensor_vector2 = np.array(input_numpy_array2,dtype=np.float64)
    torch_tesnor_vector1 = torch.from_numpy(input_tensor_vector1)
    torch_tesnor_vector2 = torch.from_numpy(input_tensor_vector2)
    torch_norm_output = torch.dot(torch.tensor(torch_tesnor_vector1),torch.tensor(torch_tesnor_vector2))
    return torch_norm_output.numpy()

input_numpy_array1 = [1.,2.,5.]
input_numpy_array2 = [3.,4.,10.]
output = dot_product_calculator(input_numpy_array1,input_numpy_array2)
print ("The dot product is:"+str(output))


The dot product is:61.0


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:9: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  if __name__ == '__main__':


The dot product is computed with the input arrays as 
input_1 = [1,2,5]
input_2 = [3,4,10]

output is = 1 * 3 + 2 * 4 + 10 * 5 = 61


In [176]:
# Example 3 - breaking (to illustrate when it breaks)
import numpy as np

def dot_product_calculator(input_numpy_array1, input_numpy_arry2,input_numpy_array3):
    input_tensor_vector1 = np.array(input_numpy_array1,dtype=np.float64)
    input_tensor_vector2 = np.array(input_numpy_array2,dtype=np.float64)
    input_tensor_vector3 = np.array(input_numpy_array3,dtype=np.float64)
    
    torch_tesnor_vector1 = torch.from_numpy(input_tensor_vector1)
    torch_tesnor_vector2 = torch.from_numpy(input_tensor_vector2)
    torch_tesnor_vector3 = torch.from_numpy(input_tensor_vector3)
    
    torch_norm_output = torch.dot(torch.tensor(torch_tesnor_vector1),torch.tensor(torch_tesnor_vector2),torch.tensor(torch_tesnor_vector3))
    return torch_norm_output.numpy()

input_numpy_array1 = [1.,2.,5.]
input_numpy_array2 = [3.,4.,10.]
input_numpy_array3 = [3.,4.,20.]

output = dot_product_calculator(input_numpy_array1,input_numpy_array2,input_numpy_array3)
print ("The dot product is:"+str(output))


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:13: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  del sys.path[0]


TypeError: dot() takes 2 positional arguments but 3 were given

Tensor.dot functions compute the dot product of two tensors. Giving an additional tensor to compute the dot product will throw an error of additional position of arguments.

#### The applications of dot product functions:

- In NLP application while computing the similarity score (i.e. cosine similarity) between the two documents we use the dot product.
- It forms a basic computation element of linear algebra. Deep learning algorithms always using dot product multiplication(i.e. for example: In calculating the output of a neuron)



## Function 3 - Torch.inverse

Computes the inverse of the square matrix input. Input can be in batches of 2D square tensors, in which case this function would return a tensor composed of individual inverses.



In [4]:
# Example 1 - working
import numpy as np

inverse_output = torch.empty(2,2)

input_matrix = [[1.,2.],[3.,4.]]
torch.inverse(torch.tensor(input_matrix), out = inverse_output)


tensor([[-2.0000,  1.0000],
        [ 1.5000, -0.5000]])

This function calculates the inverse of the square matrix which is of dimension (2,2).   

In [5]:
import numpy as np

inverse_output = torch.empty(2,2,2)

input_matrix = torch.randn(2,2,2)
print (input_matrix)
torch.inverse(torch.tensor(input_matrix), out = inverse_output)


tensor([[[ 1.7934, -1.2781],
         [ 1.1620,  0.0875]],

        [[ 0.9045,  0.3208],
         [ 0.1014, -0.5015]]])


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:7: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  import sys


tensor([[[ 0.0533,  0.7784],
         [-0.7077,  1.0922]],

        [[ 1.0316,  0.6599],
         [ 0.2085, -1.8606]]])

In this example, we randomly create the input matrices from a normal distribution. The input size is a 3-dimensional tensor for which we are calculating the inverses. The input size is (2,2,2) which generates two (2,2) matrices and it gives the inverse of those two matrices as well in a single command.

In [179]:
# Example 3 - breaking (to illustrate when it breaks)
import numpy as np

inverse_output = torch.empty(2,2,3)

input_matrix = torch.randn(2,2,3)
print (input_matrix)
torch.inverse(torch.tensor(input_matrix), out = inverse_output)


tensor([[[ 0.1338, -0.4692, -0.2628],
         [-0.7250, -2.2944,  1.1749]],

        [[ 1.0984,  0.4513, -0.3974],
         [ 0.9117, -0.2703, -0.8021]]])


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:8: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  


RuntimeError: A must be batches of square matrices, but they are 3 by 2 matrices

In this example, we randomly create the input matrices which is from a normal distribution. The input size is a 3-dimensional tensor for which we are calculating the inverses. The input size is (2,2,3) which generates two (3,2) matrices for which we cannot obtain the inverse as these are rectangular matrices.

#### Applications of inverse functions in machine learning applications:

One of the possible application of inverse of matrix is while solving linear regression in a form of matrix formulation.

- X -> be the input data (training data)
- X' -> Transpose of the input data
- Y -> be the output data (variables we are trying to predict)
- B -> Weight vector 

- The weight vector is calulated as :

<img src="Linear_regression_weight_vector_calculation.png">






## Function 4 - Torch.index_select

Returns a new tensor which indexes the input tensor along dimension dim using the entries in the index which is a LongTensor.

The returned tensor has the same number of dimensions as the original tensor (input). The dimension has the same size as the length of the index; other dimensions have the same size as in the original tensor.


In [180]:
# Example 1 - working

inverse_output = torch.empty(2,2)

input_matrix = torch.randn(2,2)
indices = torch.tensor([0])
print (input_matrix)
torch.index_select(input_matrix, 0, indices)




tensor([[ 0.0437,  1.8346],
        [-1.0896,  0.8537]])


tensor([[0.0437, 1.8346]])

In this example we randomly create the input matrices which is from a normal distribution. The input size is a (2,2) tensor from which we are extracting the first row elements using indexing operations. 

In [15]:
# Example 2 - working

input_matrix = torch.randn(2,3,4)
indices = torch.tensor([0,1])
print (input_matrix)
print ("out")
print (torch.index_select(input_matrix, 2 , indices))


tensor([[[-0.0545, -0.5610, -0.9187, -0.6235],
         [-0.5039, -0.8087,  0.4762,  1.1496],
         [-0.4608, -0.5587, -0.8092,  0.0364]],

        [[-0.8244, -0.3337,  0.1976, -0.1491],
         [ 0.3991,  0.9282,  0.2247,  0.4697],
         [-1.3364,  0.2710,  2.1608, -0.3268]]])
out
tensor([[[-0.0545, -0.5610],
         [-0.5039, -0.8087],
         [-0.4608, -0.5587]],

        [[-0.8244, -0.3337],
         [ 0.3991,  0.9282],
         [-1.3364,  0.2710]]])


In this example, we randomly create the input matrices from a normal distribution. The input is a 3 dimensional tensor of size (2,3,4) from which we are extracting the indices 0 and 1. 0 using index select function. The  `dim` parameter refers to the dimension in which we are interested to select the data with those corresponding indices. In the current example, the outermost dimension the dim value is 0, represents how many matrices are stacked vertically one over the other which is 2 in number. The `dim` value of 1 refers to the rows in those matrices. The `dim` value of 2 refers to the columns in the matrices.
From this example, we can see that passing 2 as the `dim` parameter and selecting indices as 0 and 1 we can generate the output of the resulting tensor with the first 2 columns.

In [99]:
# Example 3 - breaking (to illustrate when it breaks)
input_matrix = torch.randn(3,3,3)
indices = torch.tensor([0,4])
print (torch.index_select(input_matrix, 2 , indices))


IndexError: index out of range in self

In this example, we randomly create the input matrices which is from a normal distribution. The input size is a 3 dimensional tensor of size (2,3,3) for which we are extracting the first two column elements. But we get an index out of bounds exception as we are trying to select the 4 index which is not present as the dimension of the tensor is (3,3,3).

#### Applications of index select functions: 
Index select functions are possibly used for selecting a particular range of data to understand and get more insights about the data. This function can be used to generate required batch sizes while training the data in machine learning and deep learning applications.



## Function 5 - Cosine similarity

This function is not currently present in torch.tensor and is developed by using torch.norm and torch.dot product functions.


Cosine similarity measures the similarity between two vectors. It is measured by the cosine of the angle between two vectors and determines whether two vectors are pointing in roughly the same direction. In finding out the similarity at first we calculate the norm corresponding to the two inputs provided and then further on calculating the dot product. 

if a and b are two vectors then cosine similarity is calculated as:

$ angle = inverse(cos(dotproduct(a,b)/(norm(a)).(norm(b)))) $

Finally

cosine(angle) gives the similarity score between the two vectors.

In [16]:
# Example 1 - working

import torch
import numpy as np

def dot_product_calculator(input_numpy_array11, input_numpy_array22):
    input_tensor_vector1 = np.array(input_numpy_array11,dtype=np.float64)
    input_tensor_vector2 = np.array(input_numpy_array22,dtype=np.float64)
    torch_tesnor_vector1 = torch.from_numpy(input_tensor_vector1)
    torch_tesnor_vector2 = torch.from_numpy(input_tensor_vector2)
    torch_dot_output = torch.dot(torch.tensor(torch_tesnor_vector1),torch.tensor(torch_tesnor_vector2))
    return torch_dot_output



def vector_norm_caluclator(input_numpy_array):
    input_tensor_vector = np.array(input_numpy_array,dtype=np.float64)
    torch_tesnor_vector = torch.from_numpy(input_tensor_vector)
    torch_norm_output = torch.norm(torch.tensor(torch_tesnor_vector,dtype=torch.float))
    return torch_norm_output


def main():
    input_numpy_array1 = [1.,2.]
    input_numpy_array2 = [1.,3.]

    vectnorm_1 = vector_norm_caluclator(input_numpy_array1)
    vectnorm_2 = vector_norm_caluclator(input_numpy_array2)

    print (vectnorm_1)
    print (vectnorm_2)

    

    
    dot_product_out = dot_product_calculator(input_numpy_array1, input_numpy_array2)
    print (dot_product_out)
    dot_product_norm = torch.mul(vectnorm_1,vectnorm_2)
    
    """
    cosine similarity value and angle

    0 value  means two vectors are 90 degrees apart 
    1 value means two vectors are very close by difference is 0 degrees

    """

    print ("The score of the cosine similarity is:")
    print ((torch.div(dot_product_out,dot_product_norm)))
    print (("The angle between two vectors is:"))
    print ((torch.acos(torch.div(dot_product_out,dot_product_norm)))*57.29)



main()

tensor(2.2361)
tensor(3.1623)
tensor(7., dtype=torch.float64)
The score of the cosine similarity is:
tensor(0.9899, dtype=torch.float64)
The angle between two vectors is:
tensor(8.1293, dtype=torch.float64)


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:19: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:11: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  # This is added back by InteractiveShellApp.init_path()


We give two 1 dimensional vector with two elements as input for which we generate their respective norms and following the dot product. This is performed using torch.norm and torch.dot functions, following we calculate the cosine similarity, and finally, we use the torch.div and torch.acos function. In the above example, we use inputs as [1,2] and [1,3]. The similarity score is found to be 0.9899 and these vectors are found to be 8.1 degrees apart.

In [17]:
# Example 2 - working
# Example 1 - working
import torch
import numpy as np

def dot_product_calculator(input_numpy_array11, input_numpy_array22):
    input_tensor_vector1 = np.array(input_numpy_array11,dtype=np.float64)
    input_tensor_vector2 = np.array(input_numpy_array22,dtype=np.float64)
    torch_tesnor_vector1 = torch.from_numpy(input_tensor_vector1)
    torch_tesnor_vector2 = torch.from_numpy(input_tensor_vector2)
    torch_dot_output = torch.dot(torch.tensor(torch_tesnor_vector1),torch.tensor(torch_tesnor_vector2))
    return torch_dot_output



def vector_norm_caluclator(input_numpy_array):
    input_tensor_vector = np.array(input_numpy_array,dtype=np.float64)
    torch_tesnor_vector = torch.from_numpy(input_tensor_vector)
    torch_norm_output = torch.norm(torch.tensor(torch_tesnor_vector,dtype=torch.float))
    return torch_norm_output


def main():
    input_numpy_array1 = [1.,2.,3.]
    input_numpy_array2 = [1.,-3.,2.]

    vectnorm_1 = vector_norm_caluclator(input_numpy_array1)
    vectnorm_2 = vector_norm_caluclator(input_numpy_array2)

    print (vectnorm_1)
    print (vectnorm_2)

    

    
    dot_product_out = dot_product_calculator(input_numpy_array1, input_numpy_array2)
    
    print (dot_product_out)
    
    dot_product_norm = torch.mul(vectnorm_1,vectnorm_2)
    
    """
    cosine similarity value and angle

    0 value  means two vectors are 90 degrees apart 
    1 value means two vectors are very close by difference is 0 degrees

    """

    
    print ("The score of the cosine similarity is:")
    print ((torch.div(dot_product_out,dot_product_norm)))
    print (("The angle between two vectors is:"))
    print ((torch.acos(torch.div(dot_product_out,dot_product_norm)))*57.29)



main()

tensor(3.7417)
tensor(3.7417)
tensor(1., dtype=torch.float64)
The score of the cosine similarity is:
tensor(0.0714, dtype=torch.float64)
The angle between two vectors is:
tensor(85.8953, dtype=torch.float64)


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:19: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:11: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  # This is added back by InteractiveShellApp.init_path()


We give two 1 dimensional vector with three elements as input for which we generate their respective norms and following the dot product. This is performed using torch.norm and torch.dot functions, following we calculate the cosine similarity, and finally, we use the torch.div and torch.acos function. In the above example, we use inputs as [1,2,3] and [1,-3,2]. The similarity score is found to be 0.0714 and these vectors are found to be 85.8953 degrees apart.

In [184]:
# Example 2 - working
# Example 1 - working
import torch
import numpy as np

def dot_product_calculator(input_numpy_array11, input_numpy_array22):
    input_tensor_vector1 = np.array(input_numpy_array11,dtype=np.float64)
    input_tensor_vector2 = np.array(input_numpy_array22,dtype=np.float64)
    torch_tesnor_vector1 = torch.from_numpy(input_tensor_vector1)
    torch_tesnor_vector2 = torch.from_numpy(input_tensor_vector2)
    torch_dot_output = torch.dot(torch.tensor(torch_tesnor_vector1),torch.tensor(torch_tesnor_vector2))
    return torch_dot_output



def vector_norm_caluclator(input_numpy_array):
    input_tensor_vector = np.array(input_numpy_array,dtype=np.float64)
    torch_tesnor_vector = torch.from_numpy(input_tensor_vector)
    torch_norm_output = torch.norm(torch.tensor(torch_tesnor_vector,dtype=torch.float))
    return torch_norm_output


def main():
    input_numpy_array1 = [1.,2.,3.]
    input_numpy_array2 = [1.,-3.]

    vectnorm_1 = vector_norm_caluclator(input_numpy_array1)
    vectnorm_2 = vector_norm_caluclator(input_numpy_array2)

    print (vectnorm_1)
    print (vectnorm_2)

    

    
    dot_product_out = dot_product_calculator(input_numpy_array1, input_numpy_array2)
    
    print (dot_product_out)
    
    dot_product_norm = torch.mul(vectnorm_1,vectnorm_2)
    
    """
    cosine similarity value and angle

    0 value  means two vectors are 90 degrees apart 
    1 value means two vectors are very close by difference is 0 degrees

    """

    
    print ((torch.div(dot_product_out,dot_product_norm)))

    print ((torch.acos(torch.div(dot_product_out,dot_product_norm)))*57.29)



main()

tensor(3.7417)
tensor(3.1623)


C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:19: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
C:\Users\70725454\Anaconda3\envs\01-pytorch-basics\lib\site-packages\ipykernel_launcher.py:11: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  # This is added back by InteractiveShellApp.init_path()


RuntimeError: inconsistent tensor size, expected tensor [3] and src [2] to have the same number of elements, but got 3 and 2 elements respectively

One vector with three elements as input and another input vector with two elements are provided as inputs, we generate their respective norms and following the dot product. In the dot product, we observe that since the vector dimensions are inconsistent we will not able to calculate the cosine similarity.

#### Applications of cosine similarity functions :

In natural language processing cosine similarity can be used to find the similarity between the two documents utilizing the word document vector. 

For the feature vector in machine learning and deep learning problems if we find that two feature vectors are very similar, then we can use it to reduce the dimensionality of the feature space as well.


## Conclusion

In this notebook, we learned about the 5 basic functions of PyTorch tensor which is useful for us in different machine learning models we build in one or the other way. We understood the application and the usage of the mathematical norms which form a fundamental part of the machine learning problem. Another similar function we learned is the dot products which are directly related to the euclidean norm if all the elements are the same in the vector. Using both of these two functions norms and dot products we can find out the cosine similarity between the matrices. Finally, we tested out the inverse functions which are used for identifying the inverse of the square matrices and also we experimented with the index_select function of the torch library which would help to understand how to access the data in different dimensions of the tensor.

In the next session, I will be looking forward to building a machine learning or a deep learning model with these fundamental units of Tensors that have been learned.

## Reference Links
Provide links to your references and other interesting articles about tensors
* Official documentation for `torch.Tensor`: https://pytorch.org/docs/stable/tensors.html
* https://en.wikipedia.org/wiki/Cosine_similarity
* https://en.wikipedia.org/wiki/Matrix_norm
* https://en.wikipedia.org/wiki/Norm_(mathematics)